tests/testbed_setup/test_xr_migration.py

```diff
@@ -0,0 +1,210 @@
+    logger.debug("after mgmt config")
+
+    # XR to Sonic Migration steps
+    duthost_console.timeout = 300
+    # Copy Required Files to Harddisk from migration file location
+    duthost_console.write_and_poll(f"scp {scp_user}@{scp_host}:{mig_file_path}* /harddisk:/", "Password:")
+    duthost_console.write_channel(scp_pass + duthost_console.RETURN)
+    read_con(duthost_console, xr_prompt)
+
+    # Run XR Upgrade Check
+    duthost_console.write_channel(f"run python /harddisk:/{mig_script} --xr_upgrade --check" + duthost_console.RETURN)
+    out = read_con(duthost_console, xr_prompt)
+    if "EXIT_ON_SUCCESS" not in out:
+        duthost_console.disconnect()
+        raise Exception("Error occured in XR upgrade check, exiting.")
+
+    # Upgrade to XR Migration Image
+    duthost_console.write_channel(f"run python /harddisk:/{mig_script} --xr_upgrade" + duthost_console.RETURN)
+    out = read_con(duthost_console, "XR_UPGRADE: Device will now reload for IOS XR upgrade to")
+    duthost_console.timeout = 2000
+    out2 = read_con(duthost_console, "Press RETURN to get started.")
+    output = out + out2
+    if "[ERROR]" in output:
+        duthost_console.disconnect()
+        raise Exception("An error happened during XR upgrade, exiting")
+    duthost_console.write_and_poll(duthost_console.RETURN, "Username:")
+    duthost_console.write_and_poll(console_user, "Password:")
+    duthost_console.write_and_poll(console_password, xr_prompt)
+    logger.debug("after migration image upgrade login")
+    sleep(120)
+
+    # Install Authenticated Variable and Migrate to SONiC
+    duthost_console.write_channel(f"run python /harddisk:/{mig_script} --av_install --sonic_migration_rp" +
+                                  duthost_console.RETURN)
+
+    # Login and migrate line cards to Sonic
+    read_con(duthost_console, "sonic login:")
+    duthost_console.write_channel(console_user + duthost_console.RETURN)
+    read_con(duthost_console, "Password:")
+    duthost_console.write_channel(console_password + duthost_console.RETURN)
+    sleep(120)
+    duthost_console.write_channel(f"sudo python /mnt/obfl/{mig_script} --sonic_migration_lc" +
+                                  duthost_console.RETURN)
+    read_con(duthost_console, "user power cycle")
+    read_con(duthost_console, "sonic login:")
+    duthost_console.write_channel(console_user + duthost_console.RETURN)
+    read_con(duthost_console, "Password:")
+    duthost_console.write_channel(console_password + duthost_console.RETURN)
+    sleep(360)
+
+    # Sonic Postcheck
+    duthost_console.write_channel(f"sudo python /mnt/obfl/{mig_script} --sonic_migration_postcheck" +
+                                  duthost_console.RETURN)
+    output = read_con(duthost_console, "EXIT_ON_SUCCESS: SONiC migration script exiting due to successful completion")
+    if "[ERROR]" in output:
+        duthost_console.disconnect()
+        raise Exception("Error occured during post check, exiting")
```

Comment:
This test case could break testbed if not performed properly.
Is there any protection logic to make sure the testbed will be back with original image if it fails in the middle?

Reply:
Can we consider having recovery logic?
Or the test case is not supposed to run nightly?

If the test is supposed to be included in general nightly run, human intervention is not acceptable...

Reply:
There are too many failure scenarios for automated recovery to be feasable.

I would assume this test would not be run nightly as it is related to upgrading and downgrading equipment, not daily operation.

Reply:
Hi @Azarack , then please add logics to make sure it won't be executed in daily nightly test by default.
At the start to test, please add assertion:
1. Testbed is Cisco chassis
2. Enable this test by certain parameter, it could be --enable_testbed_setup_test

The test should be executed on when both 1 and 2 are true

Reply:
@yejianquan I have added a check to ensure the hwsku is Cisco, but I'm having trouble implementing 2.  Is there another test that does this I can look at as an example?

Reply:
Please check:
1. https://github.com/sonic-net/sonic-mgmt/blob/6132753c11522b6570c0d75be7ef7a16e855e839/tests/clock/conftest.py#L10
2. https://github.com/sonic-net/sonic-mgmt/blob/6132753c11522b6570c0d75be7ef7a16e855e839/tests/clock/conftest.py#L21

This is a good example

Reply:
Thank you for the example, I was able to update the test to check that --enable_xr_mig is passed as a cli argument and there is verification it is being run on a Cisco chassis.